### PR TITLE
Cooldown update-firmware and  flash

### PR DESCRIPTION
**Pull request by haeferer**
Hi, i started working again on my rabbit-based-version and found some actual problems: 

## patch for FastOTA Update fail on slow / instable networks (missing chunk algo)

Problem:
if your particle is connected to a slow network and you have a lot of app code, sometimes FASTOTA will fail.
 
Reason:
the missed_chunk algorithm is to aggressive (there is not enough time for the board to send missed chunk informations to the server, so the overall Update will fail)

Solution:
Slow down FastOTA Responses for missed chunks, to give particle (and the network) time to answer

### CODE REVIEW

**Comment by jlkalberer** [comment]
This is going to take a while to merge. @AntonPuko let's be very thorough with this review.

Also, how long does it take to flash the device with your change? Is there any way that we can try the very aggressive OTA first and then default to your delays after? (I haven't looked at the code yet FYI)

**Comment by haeferer** [comment]
For my tests the flashtime is comparable . You only have to wait longer for missing_chunks info (currently i tryed 500ms every 50 chunks)

Fast OTA currently works by sending Chunks as Fast as Possible and look if missed_chunks arrives during this OTA. This works fine for normal Connections but not for 
* very fast (direct WiFi) -> Particle is not fast enough to handle
* slow connections (answer from particle tooks to Long)

**Comment by jlkalberer** [comment]
Move the `.env` changes into another PR so it's easier to review. This is way too large as is.

File: src/lib/Flasher.js [hunk 1]
```diff
@@ -351,7 +359,7 @@ class Flasher {
         const messageToken = this._sendChunk(chunkIndex);
 
         // We don't need to wait for the response if using FastOTA.
-        if (!canUseFastOTA) {
+        if (canUseFastOTA) {
```
Comment (jlkalberer) [inline src/lib/Flasher.js:362]: Wouldn't this be a bug if it was `!canUseFastOTA`? Seems like this wouldn't work before...

File: src/lib/Flasher.js [hunk 2]
```diff
@@ -44,6 +44,12 @@ const CHUNK_SIZE = 256;
 const MAX_MISSED_CHUNKS = 10;
 const MAX_BINARY_SIZE = 108000; // According to the forums this is the max size for device.
 
+function waitFor(ms: number): Promise {
```
Comment (jlkalberer) [inline src/lib/Flasher.js:47]: consider:

```
const waitFor = (ms: number): Promise<void> => 
  new Promise((resolve: Promise.resolve) => setTimeout((): void => resolve(), ms));
```

File: src/lib/Flasher.js [hunk 3]
```diff
@@ -296,10 +302,12 @@ class Flasher {
         'Starting FastOTA update',
       );
     }
-
+    let chunkCount = 0;
     this._readNextChunk();
     while (this._chunk) {
       const messageToken = this._sendChunk(this._chunkIndex);
+      chunkCount += 1;
+      if (chunkCount % 50 === 0) await waitFor(500);
```
Comment (jlkalberer) [inline src/lib/Flasher.js:310]: For Photon there is generally less than 50 chunks so this code will never be run. If it does, I think 500ms is way too long of a wait time.

I think we can have small delay here of 100ms. If this is too much for the low bandwidth scenarios, the missed chunks should handle it and add a larger delay between chunks.

File: src/lib/Flasher.js [hunk 4]
```diff
@@ -351,7 +359,7 @@ class Flasher {
         const messageToken = this._sendChunk(chunkIndex);
 
         // We don't need to wait for the response if using FastOTA.
-        if (!canUseFastOTA) {
```
Comment (jlkalberer) [inline src/lib/Flasher.js:354]: Was this a bug then?

File: src/lib/Flasher.js [hunk 5]
```diff
@@ -432,16 +440,20 @@ class Flasher {
       return null;
     }
 
-    if (this._missedChunks.size) {
-      return Promise.resolve();
+    let waitCount: number = 20;
+    while (waitCount > 0) {
+      waitCount -= 1;
+      if (this._missedChunks.size > 0) {
+        logger.info(
```
Comment (jlkalberer) [inline src/lib/Flasher.js:447]: Have you actually built an environment to test this scenario?

I'm interested in the case where you have the low bandwidth. Let's say you have a failure at chunk 1 and a failure at chunk n-1. The time between sending the two chunks is over a half second. In that case this function would exit before it saw the second missed chunk and wouldn't send the second chunk.

---

I think I heard that Particle added a new event from the system firmware that may help with this scenario but I'm not 100% sure.

File: src/scripts/update-firmware-binaries.js [hunk 6]
```diff
@@ -10,7 +10,29 @@ import nullthrows from 'nullthrows';
 import { HalModuleParser } from 'binary-version-reader';
 import dotenv from 'dotenv';
 
-dotenv.config();
+function CollectDotEnv() {
```
Comment (jlkalberer) [inline src/scripts/update-firmware-binaries.js:13]: Please put this in another PR.

**Review by jlkalberer — CHANGES_REQUESTED** [review]

**Comment by jlkalberer** [comment]
I'd love to try to merge all of your other changes into this branch. I'm guessing you have done a lot of work for horizontal scaling.

File: src/lib/Flasher.js [hunk 7]
```diff
@@ -432,16 +440,20 @@ class Flasher {
       return null;
     }
 
-    if (this._missedChunks.size) {
```
Comment (jlkalberer) [inline src/lib/Flasher.js:435]: @haeferer - I think this is solved if we just remove this if-statement. We should always wait a bit for the missing chunk whether we have any or we don't.

**Pull request closed without merging**